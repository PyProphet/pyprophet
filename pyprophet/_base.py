from dataclasses import dataclass, field
import os
import copy
import sys

from loguru import logger
from .io.util import (
    is_tsv_file,
    is_sqlite_file,
    is_parquet_file,
    is_valid_single_split_parquet_dir,
    is_valid_multi_split_parquet_dir,
)


@dataclass
class BaseIOConfig:
    """
    Base configuration class for I/O-related metadata and common attributes used across algorithms.

    Attributes:
        infile (str): Path to the input file (e.g., .osw, .parquet).
        outfile (str): Path to the output file to be written.
        file_type (str): Type of the input file (e.g., 'osw', 'parquet', 'parquet_split', 'tsv').
        subsample_ratio (float): Subsampling ratio for large data.
        level (str): Scoring level (e.g., 'ms1', 'ms2', 'transition', 'alignment').
        context (str): Context or mode in which the reader/writer operates
                       (e.g., 'score_learn', 'score_apply', 'ipf', 'level_context').
        prefix (str): Automatically derived from outfile (e.g., 'results/output' for 'results/output.osw').
    """

    infile: str
    outfile: str
    file_type: str = field(init=False)
    subsample_ratio: float
    level: str
    context: str
    prefix: str = field(init=False)

    def __post_init__(self):
        """
        Initialize the file_type and prefix attributes based on the input file.
        """

        infile = self.infile

        if is_sqlite_file(infile):
            if infile.endswith(".osw"):
                self.file_type = "osw"
            elif infile.endswith(".sqmass") or infile.endswith(".sqMass"):
                self.file_type = "sqmass"
        elif is_parquet_file(infile):
            self.file_type = "parquet"
        elif is_valid_single_split_parquet_dir(infile):
            self.file_type = "parquet_split"
        elif is_valid_multi_split_parquet_dir(infile):
            self.file_type = "parquet_split_multi"
        elif is_tsv_file(infile):
            self.file_type = "tsv"
        else:
            logger.critical(
                f"Failed to infer file type for: {infile}. Supported formats are: "
                ".osw, .sqmass/.sqMass, .parquet, split parquet directories (.oswpq/oswpqd), or .tsv files.\n"
                f"  - is_sqlite_file: {is_sqlite_file(infile)}\n"
                f"  - endswith .osw: {infile.endswith('.osw')}\n"
                f"  - endswith .sqmass/.sqMass: {infile.endswith('.sqmass') or infile.endswith('.sqMass')}\n"
                f"  - is_parquet_file: {is_parquet_file(infile)}\n"
                f"  - is_valid_single_split_parquet_dir: {is_valid_single_split_parquet_dir(infile)}\n"
                f"  - is_valid_multi_split_parquet_dir: {is_valid_multi_split_parquet_dir(infile)}\n"
                f"  - is_tsv_file: {is_tsv_file(infile)}"
            )

            sys.exit(1)

        self.prefix = os.path.splitext(
            self.outfile
        )[
            0
        ]  # TODO: use pathlib instead to avoid potential cases where the outfile is a directory (split_parquet) file.oswpq/, which would result the prefix in being file.oswpq instead of file

    def __str__(self):
        return (
            f"BaseIOConfig(\ninfile='{self.infile}'\noutfile='{self.outfile}'\n"
            f"file_type='{self.file_type}'\nsubsample_ratio={self.subsample_ratio}\n"
            f"level='{self.level}'\ncontext='{self.context}'\nprefix='{self.prefix}')"
        )

    def __repr__(self):
        return (
            f"BaseIOConfig(infile='{self.infile}', outfile='{self.outfile}', "
            f"subsample_ratio={self.subsample_ratio}, level='{self.level}', "
            f"context='{self.context}')"
        )

    def copy(self):
        """
        Return a deep copy of the config object.
        """
        return copy.deepcopy(self)
