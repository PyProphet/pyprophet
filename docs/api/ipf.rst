Inference of Peptidoforms Documentation
==========================

.. automodule:: pyprophet.ipf
   :no-members:
   :no-inherited-members:

.. currentmodule:: pyprophet.ipf

.. autosummary::
   :nosignatures:
   :toctree: generated/
   :template: class.rst

   infer_peptidoforms
   peptidoform_inference
   precursor_inference
   apply_bm
   prepare_transition_bm
   transfer_confident_evidence_across_runs
   prepare_precursor_bm
   compute_model_fdr