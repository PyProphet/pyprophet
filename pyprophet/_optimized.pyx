# encoding: latin-1
cimport cython
cimport libc.stdlib
cimport numpy as np

import numpy as np
import operator


ctypedef np.float32_t DATA_TYPE


@cython.boundscheck(False)
@cython.wraparound(False)
def find_nearest_matches(DATA_TYPE[:] basis, DATA_TYPE[:] sample_points, use_sort_order=1):
    cdef size_t num_basis = basis.shape[0]
    cdef size_t num_samples = sample_points.shape[0]
    result = np.zeros((num_samples,), dtype=np.int64)
    cdef np.int64_t[:] view = result
    cdef size_t i, best_j
    cdef size_t low, high, mid
    cdef DATA_TYPE sp_i, best_dist, dist
    cdef int sort_order

    if not use_sort_order:
        for i in range(num_samples):
            sp_i = sample_points[i]
            best_j = 0
            best_dist = abs(basis[0] - sp_i)
            for j in range(1, num_basis):
                dist = abs(basis[j] - sp_i)
                if dist < best_dist:
                    best_dist = dist
                    best_j = j
            view[i] = best_j
        return result

    sort_order = find_sort_order(basis)
    for i in range(num_samples):
        sp_i = sample_points[i]
        if sort_order == 0:
            best_j = 0
            best_dist = abs(basis[0] - sp_i)
            for j in range(1, num_basis):
                dist = abs(basis[j] - sp_i)
                if dist < best_dist:
                    best_dist = dist
                    best_j = j
        elif sort_order == 1:
            low = 0
            high = num_basis - 1
            best_j = -1
            if basis[low] == sp_i:
                best_j = low
            elif basis[high] == sp_i:
                best_j = high
            else:
                while low < high - 1:
                    mid = (low + high) / 2
                    if basis[mid] == sp_i:
                        best_j = mid
                    if basis[mid] < sp_i:
                        low = mid
                    else:
                        high = mid
                if best_j == -1:
                    if abs(basis[low] - sp_i) < abs(basis[high] - sp_i):
                        best_j = low
                    else:
                        best_j = high
            # find first match in list !
            while best_j > 0:
                if basis[best_j - 1] == basis[best_j]:
                    best_j = best_j - 1
                else:
                    break
        else:
            low = 0
            high = num_basis - 1
            best_j = -1
            if basis[low] == sp_i:
                best_j = low
            elif basis[high] == sp_i:
                best_j = high
            else:
                while low < high - 1:
                    mid = (low + high) / 2
                    if basis[mid] == sp_i:
                        best_j = mid
                        break
                    if basis[mid] > sp_i:
                        low = mid
                    else:
                        high = mid
                if best_j == -1:
                    if abs(basis[low] - sp_i) < abs(basis[high] - sp_i):
                        best_j = low
                    else:
                        best_j = high
            # find first match in list:
            while best_j > 0:
                if basis[best_j - 1] == basis[best_j]:
                    best_j = best_j - 1
                else:
                    break

        view[i] = best_j
    return result


cdef int find_sort_order(DATA_TYPE[:] basis):
    # 0: unsorted
    # 1: ascending
    # 2: descending
    cdef size_t i, n
    n = basis.shape[0]
    if n <= 1:
        return 0
    i = 0
    while i < n - 1 and basis[i] == basis[i + 1]:
        i += 1
    if i == n - 1:
        return 1   # or -1 as list is constant
    if basis[i] < basis[i + 1]:
        for i in range(i, n - 1):
            if basis[i] > basis[i + 1]:
                return 0
        return 1
    else:
        for i in range(i, n - 1):
            if basis[i] < basis[i + 1]:
                return 0
        return -1


@cython.boundscheck(False)
@cython.wraparound(False)
def count_num_positives(np.float64_t[:] values):
    cdef size_t n = values.shape[0]
    cdef np.float64_t[:] inp_view = values
    cdef size_t i0, i1, c
    result = np.zeros_like(values, dtype=np.int64)
    cdef np.int64_t[:] res_view = result
    i0 = i1 = 0
    c = n
    while i0 < n:
        while i1 < n and inp_view[i0] == inp_view[i1]:
            res_view[i1] = c
            i1 += 1
        c -= 1
        i0 += 1
    return result


@cython.boundscheck(False)
@cython.wraparound(False)
def find_top_ranked(np.int64_t[:] tg_ids, DATA_TYPE[:] scores):
    cdef size_t n = scores.shape[0]
    flags = np.zeros((n,), dtype=np.int64)
    cdef np.int64_t[:] view = flags
    cdef DATA_TYPE current_max = scores[0]
    cdef size_t current_imax = 0
    cdef size_t current_tg_id = tg_ids[0]
    cdef size_t current_write_i = 0
    cdef size_t i
    cdef size_t id_
    cdef DATA_TYPE sc
    for i in range(tg_ids.shape[0]):
        id_ = tg_ids[i]
        sc = scores[i]
        if id_ != current_tg_id:
            current_tg_id = id_
            view[current_imax] = 1
            current_write_i += 1
            current_max = sc
            current_imax = i
            continue
        if sc > current_max:
            current_max = sc
            current_imax = i
    view[current_imax] = 1
    return flags

@cython.boundscheck(False)
@cython.wraparound(False)
def find_second_ranked(np.int64_t[:] tg_ids, DATA_TYPE[:] scores):
    cdef size_t n = scores.shape[0]
    flags = np.zeros((n,), dtype=np.int64)
    cdef np.int64_t[:] view = flags
    cdef DATA_TYPE current_max = scores[0]
    cdef DATA_TYPE current_2ndmax = scores[0]
    cdef size_t current_imax = 0
    cdef size_t current_i2ndmax = 0
    cdef size_t current_tg_id = tg_ids[0]
    cdef size_t current_write_i = 0
    cdef size_t i
    cdef size_t id_
    cdef DATA_TYPE sc
    for i in range(tg_ids.shape[0]):
        id_ = tg_ids[i]
        sc = scores[i]
        if id_ != current_tg_id:

            ## In case that best and second best are equal, do not select any
            ## pg (e.g. if only one pg is present)
            if current_i2ndmax != current_imax:
                view[current_i2ndmax] = 1

            # Start next group
            current_tg_id = id_
            current_write_i += 1
            current_max = sc
            current_2ndmax = sc
            current_imax = i
            current_i2ndmax = i
            continue
        if current_i2ndmax == current_imax:
            # start of group: let's initialize with the next best
            current_2ndmax = sc
            current_i2ndmax = i
        if sc > current_max:
            # Found new best group, previous best is new 2nd best
            current_2ndmax = current_max
            current_i2ndmax = current_imax
            current_max = sc
            current_imax = i
        elif sc > current_2ndmax:
            # Found new 2nd best group
            current_2ndmax = sc
            current_i2ndmax = i
    if current_i2ndmax != current_imax:
        view[current_i2ndmax] = 1
    return flags

cdef partial_rank(DATA_TYPE[:] scores, size_t imin, size_t imax, np.uint32_t[:] ranks, np.uint32_t * ix):
    """ imax is exclusive, imax-imin is the size of a target group, so 32 bit int should
    be sufficient"""
    cdef np.uint32_t i, j, pos
    for i in range(imax - imin):
        ix[i] = i
    for i in range(imax - imin - 1):
        pos = i
        for j in range(i + 1, imax - imin):
            if scores[ix[j] + imin] > scores[ix[pos] + imin]:
                pos = j
        ix[i], ix[pos] = ix[pos], ix[i]

    for j in range(imax - imin):
        ranks[ix[j] + imin] = j + 1


@cython.boundscheck(False)
@cython.wraparound(False)
def rank(np.int64_t[:] tg_ids, DATA_TYPE[:] scores):
    # we assume a transition group has always max 1000 entries:
    cdef np.uint32_t * ix = <np.uint32_t * > libc.stdlib.malloc(1000 * sizeof(np.uint32_t))
    cdef size_t n = tg_ids.shape[0]
    result = np.zeros((n,), dtype=np.uint32)
    cdef np.uint32_t[:] ranks = result
    cdef size_t imin = 0
    cdef size_t imax
    cdef np.int64_t g0
    while imin < n:
        imax = imin + 1
        g0 = tg_ids[imin]
        while imax < n and tg_ids[imax] == g0:
            imax += 1
        partial_rank(scores, imin, imax, ranks, ix)
        imin = imax
    libc.stdlib.free(ix)
    return result


@cython.boundscheck(False)
@cython.wraparound(False)
def rank32(np.uint32_t[:] tg_ids, DATA_TYPE[:] scores):
    # we assume a transition group has always max 1000 entries:
    cdef np.uint32_t * ix = <np.uint32_t * > libc.stdlib.malloc(1000 * sizeof(np.uint32_t))
    cdef size_t n = tg_ids.shape[0]
    result = np.zeros((n,), dtype=np.uint32)
    cdef np.uint32_t[:] ranks = result
    cdef size_t imin = 0
    cdef size_t imax
    cdef np.float32_t g0
    while imin < n:
        imax = imin + 1
        g0 = tg_ids[imin]
        while imax < n and tg_ids[imax] == g0:
            imax += 1
        partial_rank(scores, imin, imax, ranks, ix)
        imin = imax
    libc.stdlib.free(ix)
    return result


@cython.boundscheck(False)
@cython.wraparound(False)
def single_chromatogram_hypothesis_fast(np.float64_t[:] inv_pg_pp_true, np.float64_t prior_chrom_null, np.float64_t prior_pg_true):

    # inv_pg_pp_true contains the probabilities that the peaks are false

    cdef size_t n = inv_pg_pp_true.shape[0]
    cdef np.float64_t final_val
    cdef np.float64_t val
    cdef np.float64_t sum_all = 0.0
    result = np.zeros((n+1,), dtype=np.float64)
    cdef np.float64_t[:] cresult = result

    cdef int i
    cdef int j
    cdef double af = 1.0

    # Compute the null hypothesis (e.g. all peaks are false)
    i = 0
    while i < n:
        af *= inv_pg_pp_true[i]
        i += 1

    cresult[0] = af * prior_chrom_null

    # Compute the alternative hypotheses (e.g. one peak is true and all other
    # peaks are false)
    i = 0
    while i < n:
        val = inv_pg_pp_true[i]
        j = 0
        af = 1.0
        # Compute the product of all peaks being false except i
        while j < n:
            if i != j:
                af *= inv_pg_pp_true[j]
            j += 1

        final_val = <np.float64_t>( (1-val) * af) # value i is true

        cresult[i+1] = final_val * prior_pg_true
        i += 1

    # Compute sum over all hypothesis
    i = 0
    while i < n+1:
        sum_all += cresult[i]
        i += 1

    # Divide by sum over all hypothesis (normalize)
    i = 0
    while i < n+1:
        cresult[i] = cresult[i] / sum_all
        i += 1

    return result
