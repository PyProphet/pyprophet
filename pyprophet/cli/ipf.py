import click
from loguru import logger

from .util import write_logfile, measure_memory_usage_and_time
from .._config import IPFIOConfig
from ..ipf import infer_peptidoforms
from ..glyco.glycoform import infer_glycoforms


# IPF
@click.command()
# File handling
@click.option(
    "--in",
    "infile",
    required=True,
    type=click.Path(exists=True),
    help="PyProphet input file. Valid formats are .osw, .parquet (produced by export parquet)",
)
@click.option(
    "--out",
    "outfile",
    type=click.Path(exists=False),
    help="PyProphet output file. Valid formats are .osw, .parquet. Must be the same format as input file.",
)
# IPF parameters
@click.option(
    "--ipf_ms1_scoring/--no-ipf_ms1_scoring",
    default=True,
    show_default=True,
    help="Use MS1 precursor data for IPF.",
)
@click.option(
    "--ipf_ms2_scoring/--no-ipf_ms2_scoring",
    default=True,
    show_default=True,
    help="Use MS2 precursor data for IPF.",
)
@click.option(
    "--ipf_h0/--no-ipf_h0",
    default=True,
    show_default=True,
    help="Include possibility that peak groups are not covered by peptidoform space.",
)
@click.option(
    "--ipf_grouped_fdr/--no-ipf_grouped_fdr",
    default=False,
    show_default=True,
    help="[Experimental] Compute grouped FDR instead of pooled FDR to better support data where peak groups are evaluated to originate from very heterogeneous numbers of peptidoforms.",
)
@click.option(
    "--ipf_max_precursor_pep",
    default=0.7,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored precursors in IPF.",
)
@click.option(
    "--ipf_max_peakgroup_pep",
    default=0.7,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored peak groups in IPF.",
)
@click.option(
    "--ipf_max_precursor_peakgroup_pep",
    default=0.4,
    show_default=True,
    type=float,
    help="Maximum BHM layer 1 integrated precursor peakgroup PEP to consider in IPF.",
)
@click.option(
    "--ipf_max_transition_pep",
    default=0.6,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored transitions in IPF.",
)
@click.option(
    "--propagate_signal_across_runs/--no-propagate_signal_across_runs",
    default=False,
    show_default=True,
    help="Propagate signal across runs (requires running alignment).",
)
@click.option(
    "--ipf_max_alignment_pep",
    default=1.0,
    show_default=True,
    type=float,
    help="Maximum PEP to consider for good alignments.",
)
@click.option(
    "--across_run_confidence_threshold",
    default=0.5,
    show_default=True,
    type=float,
    help="Maximum PEP to consider for propagating signal across runs for aligned features.",
)
@click.pass_context
@measure_memory_usage_and_time
@logger.catch(reraise=True)
def ipf(
    ctx,
    infile,
    outfile,
    ipf_ms1_scoring,
    ipf_ms2_scoring,
    ipf_h0,
    ipf_grouped_fdr,
    ipf_max_precursor_pep,
    ipf_max_peakgroup_pep,
    ipf_max_precursor_peakgroup_pep,
    ipf_max_transition_pep,
    propagate_signal_across_runs,
    ipf_max_alignment_pep,
    across_run_confidence_threshold,
):
    """
    Infer peptidoforms after scoring of MS1, MS2 and transition-level data.
    """

    # Check to see if we're calling from the previous deprecated command call
    if ctx.command.deprecated:
        logger.warning(
            "The 'pyprophet ipf' command is deprecated and will be removed in future versions. Use 'pyprophet infer peptidoform' instead."
        )

    if outfile is None:
        outfile = infile
    else:
        outfile = outfile

    config = IPFIOConfig.from_cli_args(
        infile,
        outfile,
        1,  # Subsample ratio is not applicable for IPF
        "ipf",  # Level is not applicable for IPF
        "ipf",
        ipf_ms1_scoring,
        ipf_ms2_scoring,
        ipf_h0,
        ipf_grouped_fdr,
        ipf_max_precursor_pep,
        ipf_max_peakgroup_pep,
        ipf_max_precursor_peakgroup_pep,
        ipf_max_transition_pep,
        propagate_signal_across_runs,
        ipf_max_alignment_pep,
        across_run_confidence_threshold,
    )
    write_logfile(
        ctx.obj["LOG_LEVEL"], f"{config.prefix}_pyp_ipf.log", ctx.obj["LOG_HEADER"]
    )
    infer_peptidoforms(config)


# Infer glycoforms
@click.command()
@click.option(
    "--in", "infile", required=True, type=click.Path(exists=True), help="Input file."
)
@click.option("--out", "outfile", type=click.Path(exists=False), help="Output file.")
@click.option(
    "--ms1_precursor_scoring/--no-ms1_precursor_scoring",
    default=True,
    show_default=True,
    help="Use MS1 precursor data for glycoform inference.",
)
@click.option(
    "--ms2_precursor_scoring/--no-ms2_precursor_scoring",
    default=True,
    show_default=True,
    help="Use MS2 precursor data for glycoform inference.",
)
@click.option(
    "--grouped_fdr/--no-grouped_fdr",
    default=False,
    show_default=True,
    help="[Experimental] Compute grouped FDR instead of pooled FDR to better support data where peak groups are evaluated to originate from very heterogeneous numbers of glycoforms.",
)
@click.option(
    "--max_precursor_pep",
    default=1,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored precursors.",
)
@click.option(
    "--max_peakgroup_pep",
    default=0.7,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored peak groups.",
)
@click.option(
    "--max_precursor_peakgroup_pep",
    default=1,
    show_default=True,
    type=float,
    help="Maximum BHM layer 1 integrated precursor peakgroup PEP to consider.",
)
@click.option(
    "--max_transition_pep",
    default=0.6,
    show_default=True,
    type=float,
    help="Maximum PEP to consider scored transitions.",
)
@click.option(
    "--use_glycan_composition/--use_glycan_struct",
    "use_glycan_composition",
    default=True,
    show_default=True,
    help="Compute glycoform-level FDR based on glycan composition or struct.",
)
@click.option(
    "--ms1_mz_window",
    default=10,
    show_default=True,
    type=float,
    help="MS1 m/z window in Thomson or ppm.",
)
@click.option(
    "--ms1_mz_window_unit",
    default="ppm",
    show_default=True,
    type=click.Choice(["ppm", "Da", "Th"]),
    help="MS1 m/z window unit.",
)
@click.option(
    "--propagate_signal_across_runs/--no-propagate_signal_across_runs",
    default=False,
    show_default=True,
    help="Propagate signal across runs (requires running alignment).",
)
@click.option(
    "--max_alignment_pep",
    default=1.0,
    show_default=True,
    type=float,
    help="Maximum PEP to consider for good alignments.",
)
@click.option(
    "--across_run_confidence_threshold",
    default=0.5,
    show_default=True,
    type=float,
    help="Maximum PEP to consider for propagating signal across runs for aligned features.",
)
def glycoform(
    infile,
    outfile,
    ms1_precursor_scoring,
    ms2_precursor_scoring,
    grouped_fdr,
    max_precursor_pep,
    max_peakgroup_pep,
    max_precursor_peakgroup_pep,
    max_transition_pep,
    use_glycan_composition,
    ms1_mz_window,
    ms1_mz_window_unit,
    propagate_signal_across_runs,
    max_alignment_pep,
    across_run_confidence_threshold,
):
    """
    Infer glycoforms after scoring of MS1, MS2 and transition-level data.
    """

    if outfile is None:
        outfile = infile

    infer_glycoforms(
        infile=infile,
        outfile=outfile,
        ms1_precursor_scoring=ms1_precursor_scoring,
        ms2_precursor_scoring=ms2_precursor_scoring,
        grouped_fdr=grouped_fdr,
        max_precursor_pep=max_precursor_pep,
        max_peakgroup_pep=max_peakgroup_pep,
        max_precursor_peakgroup_pep=max_precursor_peakgroup_pep,
        max_transition_pep=max_transition_pep,
        use_glycan_composition=use_glycan_composition,
        ms1_mz_window=ms1_mz_window,
        ms1_mz_window_unit=ms1_mz_window_unit,
        propagate_signal_across_runs=propagate_signal_across_runs,
        max_alignment_pep=max_alignment_pep,
        across_run_confidence_threshold=across_run_confidence_threshold,
    )
