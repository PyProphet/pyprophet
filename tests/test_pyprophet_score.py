from __future__ import print_function

import os
import subprocess
import shutil
import sys

import pandas as pd
import sqlite3

import pytest

from pyprophet.ipf import read_pyp_peakgroup_precursor

pd.options.display.expand_frame_repr = False
pd.options.display.precision = 4
pd.options.display.max_columns = None

DATA_FOLDER = os.path.join(os.path.dirname(os.path.abspath(__file__)), "data")

def _run_cmdline(cmdline):
    stdout = cmdline + "\n"
    try:
        stdout += str(subprocess.check_output(cmdline, shell=True,
                                          stderr=subprocess.STDOUT))
    except subprocess.CalledProcessError as error:
        print(error, end="", file=sys.stderr)
        raise
    return stdout

def _run_pyprophet_tsv_to_learn_model(regtest, temp_folder, dump_result_files=False, parametric=False, pfdr=False, pi0_lambda=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.txt")
    shutil.copy(data_path, temp_folder)
    cmdline = "pyprophet score --pi0_method=smoother --pi0_lambda 0.4 0 0 --in=test_data.txt --test --ss_iteration_fdr=0.02"
    if parametric:
        cmdline += " --parametric"
    if pfdr:
        cmdline += " --pfdr"
    if pi0_lambda is not False:
        cmdline += " --pi0_lambda=" + pi0_lambda
    stdout = _run_cmdline(cmdline)

    print(pd.read_csv("test_data_summary_stat.csv", sep=",", nrows=100).sort_index(axis=1),file=regtest)
    print(pd.read_csv("test_data_full_stat.csv", sep=",", nrows=100).sort_index(axis=1),file=regtest)
    print(pd.read_csv("test_data_scored.tsv", sep="\t", nrows=100).sort_index(axis=1),file=regtest)
    print(pd.read_csv("test_data_weights.csv", sep=",", nrows=100).sort_index(axis=1),file=regtest)

def _run_pyprophet_osw_to_learn_model_metabo(regtest, temp_folder, dump_result_files=False, ms1ms2=False, score_filter=False):
    os.chdir(temp_folder)
    data_path_metabo = os.path.join(DATA_FOLDER, "test_data_compound.osw")
    shutil.copy(data_path_metabo, temp_folder)

    # MS1-level
    cmdline = "pyprophet score --in=test_data_compound.osw --level=ms1 --test --ss_iteration_fdr=0.02"
    if score_filter:
        cmdline += " --ss_score_filter var_isotope_overlap_score,var_massdev_score,var_xcorr_coelution,var_isotope_correlation_score"

    # MS2-level
    if ms1ms2:
            cmdline += " score --in=test_data_compound.osw --level=ms1ms2 --test --ss_iteration_fdr=0.02"
            if score_filter:
                cmdline += " --ss_score_filter metabolomics"
    else:
        cmdline += " score --in=test_data_compound.osw --level=ms2 --test --ss_iteration_fdr=0.02"
        if score_filter:
            cmdline += " --ss_score_filter var_isotope_overlap_score,var_isotope_correlation_score,var_intensity_score,var_massdev_score,var_library_corr,var_norm_rt_score"

    stdout = _run_cmdline(cmdline)

    scored = sqlite3.connect("test_data_compound.osw")
    table = pd.read_sql_query("SELECT * FROM PYPROPHET_WEIGHTS ORDER BY score", scored)
    scored.close()

    print(table.head(100).sort_index(axis=1),file=regtest)

def _run_pyprophet_osw_to_learn_model(regtest, temp_folder, dump_result_files=False, parametric=False, pfdr=False, pi0_lambda=False, ms1ms2=False, xgboost=False, xgboost_tune=False, score_filter=False):
    os.chdir(temp_folder)
    data_path = os.path.join(DATA_FOLDER, "test_data.osw")
    shutil.copy(data_path, temp_folder)

    # MS1-level
    cmdline = "pyprophet score --in=test_data.osw --level=ms1 --test --ss_iteration_fdr=0.02"
    if parametric:
        cmdline += " --parametric"
    if pfdr:
        cmdline += " --pfdr"
    if pi0_lambda is not False:
        cmdline += " --pi0_lambda=" + pi0_lambda
    if xgboost:
        cmdline += " --classifier=XGBoost"
    if xgboost_tune:
        cmdline += " --xgb_autotune"
    if score_filter:
        cmdline += " --ss_score_filter var_isotope_overlap_score,var_massdev_score,var_xcorr_coelution,var_isotope_correlation_score"

    # MS2-level
    if ms1ms2:
        cmdline += " score --in=test_data.osw --level=ms1ms2 --test --ss_iteration_fdr=0.02"
        if score_filter:
            cmdline += " --ss_score_filter var_ms1_isotope_overlap_score,var_ms1_massdev_score,var_ms1_xcorr_coelution,var_ms1_isotope_correlation_score,var_isotope_overlap_score,var_isotope_correlation_score,var_intensity_score,var_massdev_score,var_library_corr,var_norm_rt_score"
    else:
        cmdline += " score --in=test_data.osw --level=ms2 --test --ss_iteration_fdr=0.02"
        if score_filter:
            cmdline += " --ss_score_filter var_isotope_overlap_score,var_isotope_correlation_score,var_intensity_score,var_massdev_score,var_library_corr,var_norm_rt_score"
    if parametric:
        cmdline += " --parametric"
    if pfdr:
        cmdline += " --pfdr"
    if pi0_lambda is not False:
        cmdline += " --pi0_lambda=" + pi0_lambda
    if xgboost:
        cmdline += " --classifier=XGBoost"
    if xgboost_tune:
        cmdline += " --xgb_autotune"

    # transition-level
    cmdline += " score --in=test_data.osw --level=transition --test --ss_iteration_fdr=0.02"

    if parametric:
        cmdline += " --parametric"
    if pfdr:
        cmdline += " --pfdr"
    if pi0_lambda is not False:
        cmdline += " --pi0_lambda=" + pi0_lambda
    if xgboost:
        cmdline += " --classifier=XGBoost"
    if xgboost_tune:
        cmdline += " --xgb_autotune"

    stdout = _run_cmdline(cmdline)

    table = read_pyp_peakgroup_precursor("test_data.osw", 1.0, True, True)

    print(table.head(100).sort_index(axis=1),file=regtest)

def test_tsv_0(tmpdir, regtest):
    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True)

def test_tsv_1(tmpdir, regtest):
    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True, True)

def test_tsv_2(tmpdir, regtest):
    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True, False, True)

def test_tsv_3(tmpdir, regtest):
    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True, False, False, "0.3 0.55 0.05")

def test_tsv_apply_weights(tmpdir, regtest):

    _run_pyprophet_tsv_to_learn_model(regtest, tmpdir.strpath, True)

    _run_cmdline("pyprophet score --pi0_method=smoother --pi0_lambda 0.4 0 0 --in=test_data.txt --apply_weights=test_data_weights.csv "
                          "--test --ss_iteration_fdr=0.02")

def test_osw_0(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, pi0_lambda="0 0 0")

def test_osw_1(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, True, pi0_lambda="0 0 0")

def test_osw_2(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0")

def test_osw_3(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", ms1ms2=True)

def test_osw_4(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", ms1ms2=True, xgboost=True)

def test_osw_5(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", ms1ms2=True, xgboost=True, xgboost_tune=True)

def test_osw_6(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", score_filter=True)

def test_osw_7(tmpdir, regtest):
    _run_pyprophet_osw_to_learn_model(regtest, tmpdir.strpath, True, False, True, pi0_lambda="0 0 0", ms1ms2=True, score_filter=True)

def test_osw_8(tmpdir, regtest):
  _run_pyprophet_osw_to_learn_model_metabo(regtest, tmpdir.strpath, True, ms1ms2=False, score_filter=False)

def test_osw_9(tmpdir, regtest):
  _run_pyprophet_osw_to_learn_model_metabo(regtest, tmpdir.strpath, True, ms1ms2=False, score_filter=True)

def test_osw_10(tmpdir,regtest):
  _run_pyprophet_osw_to_learn_model_metabo(regtest, tmpdir.strpath, True, ms1ms2=True, score_filter=True)

def test_not_unique_tg_id_blocks(tmpdir):

    os.chdir(tmpdir.strpath)
    data_path = os.path.join(DATA_FOLDER, "test_invalid_data.txt")
    shutil.copy(data_path, tmpdir.strpath)
    cmdline = "pyprophet score --pi0_method=smoother --pi0_lambda 0.4 0 0 --in=test_invalid_data.txt --test --ss_iteration_fdr=0.02"

    with pytest.raises(subprocess.CalledProcessError) as exc_info:
        subprocess.check_output(cmdline, shell=True, stderr=subprocess.STDOUT)

    e = exc_info.value
    assert "Error: group_id values do not form unique blocks in input file(s)." in str(e.output)
