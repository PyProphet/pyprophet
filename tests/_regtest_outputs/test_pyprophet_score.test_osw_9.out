   level                          score   weight
0    ms1           main_var_xcorr_shape   3.6907
1    ms2           main_var_xcorr_shape   1.8019
2    ms2            var_intensity_score   2.9725
3    ms1  var_isotope_correlation_score   0.0285
4    ms2  var_isotope_correlation_score   0.7824
5    ms1      var_isotope_overlap_score  -0.0581
6    ms2      var_isotope_overlap_score  -0.3541
7    ms2               var_library_corr   0.9053
8    ms1              var_massdev_score   0.0211
9    ms2              var_massdev_score  -0.0305
10   ms2              var_norm_rt_score -60.9504
11   ms1            var_xcorr_coelution  -0.0515
